File: ruby/bob/bob.rb
class Bob
  SHOUTING = lambda {|input| input == input.upcase}

  EMPTY = lambda {|input| input.strip.empty?}

  QUESTION = lambda {|input| input.end_with? '?'}


  def hey(phrase)
    phrase ||= ''

    case phrase
    when EMPTY    then 'Fine. Be that way!'
    when SHOUTING then 'Woah, chill out!'
    when QUESTION then 'Sure.'
    else
      'Whatever.'
    end
  end

end

File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.Char

isSilence = all isSpace
isYelling s = (any isUpper s) && (not $ any isLower s)
isQuestion = (== '?') . last

responseFor s
  | isSilence s  = "Fine. Be that way!"
  | isYelling s  = "Woah, chill out!"
  | isQuestion s = "Sure."
  | otherwise    = "Whatever."
